### DISCUSSION TITLE
How to use sidecar to run tauri with a console app?

### DISCUSSION THREAD

**Discussion by Hellager** [tauri-apps/tauri#5863] [Q&A] [answered]
I'm trying figure out how to properly use sidecar to run tauri with a console app building by c#.
Here's my test project [tauri-sidecar-test](https://github.com/Hellager/tauri-sidecar-test).  Run the console app with option `-l` will print some language code in console like `en-US`, `zh-CN`, try get these language code in tauri app.
Follow the guidance [Embedding External Binaries](https://tauri.app/v1/guides/building/sidecar/), properly set `tauri.conf.json`.
If try execute sidecar in frontent, `console.log(output)` will have this error message `Uncaught (in promise) invalid utf-8 sequence of 1 bytes from index 2`. Since i've already set the console encoding to UTF8, i'm not sure why it fails.
```typescript
  async function greet() {
    // Learn more about Tauri commands at https://tauri.app/v1/guides/features/command
    setGreetMsg(await invoke("greet", { name }));

    const command = Command.sidecar("console/MyConsoleApp", ["-l"])
    const output = await command.execute();
    console.log(output); // Get `Uncaught (in promise) invalid utf-8 sequence of 1 bytes from index 2`
  }
```
Then i try execute sidecar in rust, and it will be four times `Failed to execute sidecar`. (The number of language codes is 5.)
```rust
    let (mut rx, _) = Command::new_sidecar("MyConsoleApp")
    .expect("failed to create `MyConsoleApp` binary command")
    .args(["-l"])
    .spawn()
    .expect("Failed to spawn sidecar");

    tauri::async_runtime::spawn(async move {
        while let Some(event) = rx.recv().await {
            if let CommandEvent::Stdout(message) = event {
                println!("{}", message);
            } else {
                println!("Failed to execute sidecar");
            }
        }
    });
```
So am i did anything wrong? How to properly using sidecar with a console app?

**Comment by akaigyouhou** [comment]
I guess you should use some like new_sidecar(...).encoding(..).spawn(..), but I am still quite not sure the correct answer

**Reply by Hellager** [reply]
Thanks for you help, i just forget to update this discussion.
The problem is the console app cannot correctly load the dll specific in tauri->bundle->resources.
This may be fixed soon with this pr [feat: allow specifying a resource map, closes #5844 #5950](https://github.com/tauri-apps/tauri/pull/5950)

**Comment by Hellager** [comment, marked as answer]
I have asked an issue [[bug] Always get Uncaught (in promise) invalid utf-8 sequence of 1 bytes from index 2 with sidecar](https://github.com/tauri-apps/tauri/issues/5912) and figured out why.
The problem is the console app cannot correctly load the dll specific in tauri->bundle->resources.
So the solution is quite simply, bundle all your assemblies like dll to your console app will fix it.
If you are using c# .net framework to build your console app and have the same issue, strongly recommend to use [Costura.Fody](https://github.com/Fody/Costura). For more detail, check this [Embedding DLLs in a compiled executable](https://stackoverflow.com/questions/189549/embedding-dlls-in-a-compiled-executable)
